  Shubh Agarwal

AIML B2

22070126108

# LOADING DATASET

In [37]:
import numpy as np
from sklearn.datasets import load_iris
from scipy.spatial import distance
import pandas as pd

In [ ]:
df = load_iris()

In [ ]:
df

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [ ]:
x = df.data
y = df.target


In [ ]:
x.shape

(150, 4)

In [ ]:
y.shape

(150,)

In [ ]:
y =   y.reshape(-1,1)

In [ ]:
y.shape

(150, 1)

In [33]:
# Selecting two samples for demonstration
sample1 =x[0]
sample2 = x[100]

In [34]:
sample1

array([5.1, 3.5, 1.4, 0.2])

In [35]:
sample2

array([6.3, 3.3, 6. , 2.5])

# Euclidean Distance

Euclidean distance is the distance between two points in n dimensional space

In [38]:
euc_distance = distance.euclidean(sample1,sample2)


In [39]:
euc_distance

5.2848841046895245

**Inference**




*   It shows that the distance between the two points is quite high showing the disimlarity in the features space.
*   The Degree of disimlarity is high

*   Euclidean distance creates a straight line between the point x[0] and x[100] using the fromula sqrt(x[0] - x[100])^2 . These points are in same dimensions.
*   It also shows that these points cannot be clustered together in the same class as they high degree of disimilarity .





# Manhattan Distance

Computes the distance as the sum of absolute differences between the coordinates of the points.

In [42]:
manhattan_distance = distance.cityblock(sample1,sample2)

In [43]:
manhattan_distance

8.3

**Inferences**



*  In the context of the Iris dataset, where each sample is described by four features (sepal length, sepal width, petal length, and petal width), the Manhattan distance of 8.3 suggests a significant cumulative difference across these features between the two samples.

*   Manhattan distance is calculated as the sum of the absolute differences between the coordinates of the points whoch comes out to be 8.3 for sample x[0] and x[100] (sampele1 and sample2). Unlike Euclidean distance, which considers the straight-line distance, Manhattan distance measures the distance traveled along the grid-like paths. Therefore, the Manhattan distance is independent of direction.
* The Manhattan distance between two locations is 8.3 units, it implies that the shortest path along the city grid requires traversing approximately 8 blocks in one direction and 3 blocks in the perpendicular direction (or vice versa), totaling 8.3 blocks.


*   Manhattan distance generally focus on individual features . Since Manhattan distance is greater then the euclidean distance it suggests that the difference between the two samples is greately due to the individual features rather than the overall trends or patterns



# Minkowski distance



*   The Minkowski distance is a generalized distance metric that encompasses both Euclidean and Manhattan distances as special cases.
*  Formula for calulation minkowiski distance is (( |sample1 - sample100| )^p)^1/p.


*   when p = 2 , it reduces to the Euclidean distance.
*   when p = 1 , it reduces to the manhattan distance.










In [52]:
minkowiski_distance = distance.minkowski(sample1, sample2, p=3)

In [53]:
minkowiski_distance

4.8093423374296735



*   In this case, with p = 3 ,the Minkowski distance represents a compromise between the Euclidean and Manhattan distances, offering a balanced perspective on the dissimilarity between the samples.
*   Compared to the Euclidean and Manhattan distances, the Minkowski distance with p=3 , places more emphasis on larger feature differences while still considering the overall patterns in the data.

*   A Minkowski distance of 4.81 suggests significant differences in individual feature values between the two samples, with certain features contributing more to the dissimilarity than others.
*   When the Minkowski distance is significantly larger than the Euclidean distance but smaller than the Manhattan distance, it suggests a balance between considering overall patterns and individual feature differences.This scenario implies that the dissimilarity between the samples is influenced by both the overall patterns in the data and the differences in individual feature values, but not to the extreme extent observed in the Manhattan distance.


*   This distance value can inform various analytical tasks such as clustering, classification, or similarity analysis, where understanding the dissimilarity between samples is crucial which suggests that they cannot be clustered in same class.








**Value of p**



*   As
p increases, the emphasis on larger feature differences between samples also increases.
Higher values of
p amplify the effect of larger differences in individual feature values on the overall Minkowski distance.
Consequently, the distance metric becomes more sensitive to outliers or extreme feature values.
* Small values of (e.g.,
p=1) tend to focus more on individual feature differences, while larger values of
p (e.g.
p→∞) prioritize overall patterns in the data.



# JACCARD DISTANCE

In [59]:
# Boolean-valued vectors for Jaccard Distance
vector1 = [True, False, False, True]
vector2 = [False, True, False, True]

In [60]:
from sklearn.metrics import jaccard_score

In [65]:
jaccard_score(vector1, vector2)

0.3333333333333333

In [66]:
# Calculate Jaccard Distance
jaccard_distance = 1- jaccard_score(vector1, vector2)

In [67]:
jaccard_distance

0.6666666666666667

**Inference**



*   The Jaccard Distance is a metric used to measure dissimilarity between sets. In this case, the sets correspond to the presence or absence of values (True or False) in the Boolean vectors.
*   A Jaccard Distance of 0 indicates that the sets are identical, while a value of 1 indicates that the sets have no elements in common.
A Jaccard Distance between 0 and 1 indicates the proportion of elements that are different between the sets.

*  In this scenario, with a Jaccard Distance of approximately 0.67, it suggests that about 67% of the elements between the two Boolean vectors are different.
*   Understanding such dissimilarities is crucial in various applications like text analysis (e.g., document similarity), recommendation systems, or even genetic analysis.




# Conclusion

The data exhibits varied dissimilarity characteristics: Euclidean captures overall trends, Manhattan emphasizes feature differences, Minkowski offers a balance, while Jaccard quantifies set dissimilarity. Understanding these metrics aids in nuanced analysis, crucial for diverse data-driven tasks.